src/os_net/os_net.c

```diff
@@ -40,64 +40,58 @@ static socklen_t us_l = sizeof(n_us);
 
 
 /* Bind a specific port */
-static int OS_Bindport(u_int16_t _port, unsigned int _proto, const char *_ip, int ipv6)
+int OS_Bindport(char *_port, unsigned int _proto, const char *_ip)
 {
-    int ossock;
-    struct sockaddr_in server;
+    int ossock = 0, s;
+    struct addrinfo hints, *result, *rp;
 
-#ifndef WIN32
-    struct sockaddr_in6 server6;
-#else
-    ipv6 = 0;
-#endif
 
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_INET6;    /* Allow IPv4 or IPv6 */
+    hints.ai_protocol = _proto;
     if (_proto == IPPROTO_UDP) {
-        if ((ossock = socket(ipv6 == 1 ? PF_INET6 : PF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
-            return OS_SOCKTERR;
-        }
+        hints.ai_socktype = SOCK_DGRAM;
     } else if (_proto == IPPROTO_TCP) {
-        int flag = 1;
-        if ((ossock = socket(ipv6 == 1 ? PF_INET6 : PF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
-            return (int)(OS_SOCKTERR);
-        }
-
-        if (setsockopt(ossock, SOL_SOCKET, SO_REUSEADDR,
-                       (char *)&flag,  sizeof(flag)) < 0) {
-            OS_CloseSocket(ossock);
-            return (OS_SOCKTERR);
-        }
+        hints.ai_socktype = SOCK_STREAM;
     } else {
-        return (OS_INVALID);
+        return(OS_INVALID);
     }
+    hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG | AI_V4MAPPED;
```

Comment:
On Sat, 28 Mar 2015, ddpbsd wrote:

>>      }
>> +    hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG | AI_V4MAPPED;
>
> If we're going to intentionally not supportmajor features on openbsd, why
> not drop support for it? I wonder how other projecta manage to support both
> ipv6 and ipv4 on openbsd.

They probably support one or the other but not both simultaneously as a 
server/listener app.  There are probably a few that go through the trouble 
of running multiple sockets with all the additional complexity.  If the 
application already has a need to flexibly control which addresses or 
interfaces it's listening on, then the multi-socket listening code 
probably already exists.  But the majority of server apps really aren't 
designed to know or explicitly specify individual addresses or interfaces 
they listen on.  Most of them were probably running as single-socket 
servers before IPv6 came along and probably still are after they were 
updated for IPv6.

If it's really important to have this feature in OSSEC server, I suppose 
it can be added in at some point.  But doing so just to fix a shortcoming 
in an OS that intentionally breaks RFC compliance without providing a 
configuration knob - I'd rather spend time getting basic IPv6 funtionality 
working first.

OpenBSD can continue to be used to run OSSEC server with the caveat that 
it can accept agent connections over IPv4 only and leave it at that.

Antonio Querubin
e-mail:  tony@lavanauts.org
xmpp:  antonioquerubin@gmail.com